custom_model_runner/datarobot_drum/drum/main.py

```diff
@@ -92,11 +92,11 @@ def signal_handler(sig, frame):
             try:
                 loader = RuntimeParametersLoader(options.runtime_params_file, options.code_dir)
                 loader.setup_environment_variables()
-                if RuntimeParameters.has("CUSTOM_MODEL_WORKERS"):
-                    options.max_workers = RuntimeParameters.get("CUSTOM_MODEL_WORKERS")
             except RuntimeParameterException as exc:
                 print(str(exc))
                 exit(255)
+        if RuntimeParameters.has("CUSTOM_MODEL_WORKERS"):
```

Comment:
this seems _very_ iffy.  This tends to indicate that the workers is a special case and isn't actually a runtime parameter at all.  that kind of makes sense.  These runtime parameters are user-injected env vars and user-inject secret env vars.  they aren't about configuration of how the server is run and so this looks very much like the number of workers here is not really a runtime param, but a regular param that gets passed in to the `drum server` command using the args parser (these can also be passed in as env vars that get picked up by that args parser).  If you need some help understanding how the args parsing works, let me know.

Reply:
Just want to make it consistent for end user. As user define runtime parameters CUSTOM_MODEL_WORKERS via UI in DataRobot  it can be not automatically used by drum  as it could be started without runtime_params_file option. There is a way to manually get CUSTOM_MODEL_WORKERS value from env vars and  use it when start drum with   --max-workers  option  but it looks to me like workaround.

Reply:
can you provide links to the DR code, please?  I'm missing the context here.

Reply:
I've discovered it on staging  when try to define CUSTOM_MODEL_WORKERS  runtime parameter and my custom start_server.sh run  drum without  runtime_params_file option then CUSTOM_MODEL_WORKERS is not used. As I added it only for use case when user use runtime_params_file option when start drum

Reply:
looks like it is here https://github.com/datarobot/DataRobot/blob/19d867ee7e6af21295b5d68d98ba0bce58e7a732/raptor_lake/custom_model/services/executable_service.py#L992

Reply:
but why is the custom model workers using launch config _at all_?  it's not a launch config.  it's a setting.

Reply:
since options.max_workers was already in place to  controls a number of processes in the drum server. And at the same time we had a requirement that we would like to introduce runtime parameter CUSTOM_MODEL_WORKERS which could be used to controls a number of processes  and Users can override the runtime parameter at deployment time.

Reply:
So as I understand, RuntimeParameters are intended to be only used within custom.py, while here we use it directly in DRUM.
If we want to pass some values to DRUM, the right way to do this, probably would be to add an api  drum.set_number_of_workers() and call it in the `init()` hook. Or smth like that. (If I remember correct, server is started after init() and load_model() hooks)

Does it sound about right?
@eric-s-s ?

Reply:
This should probably be part of the resource allocations that are set as part of a custom model version like this 
![image](https://github.com/user-attachments/assets/9f8bf933-0c66-4e37-8704-1f92c9043fc2)

and not set as runtime parameters.

Reply:
@yura-ant this is where we define env vars that are `drum server` settings https://github.com/datarobot/DataRobot/blob/19d867ee7e6af21295b5d68d98ba0bce58e7a732/raptor_lake/custom_model/services/executable_service.py#L919   .  The drum server is read from the env var like this test shows: https://github.com/datarobot/datarobot-user-models/pull/1126

Reply:
> I am looking for the simplest option for end user to avoid update multiple files before start using CUSTOM_MODEL_WORKERS runtime parameters.

If you say that there is a requirement to make CUSTOM_MODEL_WORKERS user defined, we should brainstorm together on how to do this. Maybe we could define system RuntimeParams and reserve some names, and then properly set env vars in the: `setup_environment_variables`
I suggest to invite @elatt and @ahamino to the discussion

Reply:
We don't really have a scheme for "system" defined runtime params. We had some brainstorming sessions in the past on adding "execution environment defined" runtime params that a custom model would inherit based on what execution environment was used. This is not a trivial thing to add though and the work has not been prioritized.

I don't know if I totally agree that this belongs with the replica and mem settings because those are k8s scheduling related stuff so the DR platform has control of them. This option is really specific to DRUM so it does seem more like it should be associated with the environment (I.e. envs with older versions of drum don't support it).

However in lieu of actually doing the big work to add exec env defined parameters, we have opted to just allow them to be defined lazily (only if a user wants to actually override them). Since this is becoming a common pattern (GPU envs, GenAI envs, and now this) we should improve the interface in drum for dealing with undefined/optional params.

Reply:
Just want to add that implementing this using a runtime parameter was in [the spec](https://docs.google.com/document/d/1yq-T7-pmjW49RsvbslNIl3BF_GRbYLti2McVmZgd_To/edit#heading=h.dajgx8borbez), so it would be been great to have this conversation earlier.

I agree, that using runtime parameters for this kind of configuration is not optimal but it fits the requirements without having to do a lot of extra work.

Reply:
> @yura-ant this is where we define env vars that are `drum server` settings https://github.com/datarobot/DataRobot/blob/19d867ee7e6af21295b5d68d98ba0bce58e7a732/raptor_lake/custom_model/services/executable_service.py#L919 . The drum server is read from the env var like this test shows: #1126

I can try to use MAX_WORKER  in  executable_service.py if CUSTOM_MODEL_WORKERS runtime parameter was defined. But I still think that current PR is better option today as it will allow use  CUSTOM_MODEL_WORKERS  runtime parameter without DataRobot app in case when it was defined without runtime_params_file. Anyway we need to sync CUSTOM_MODEL_WORKERS to MAX_WORKER  if  runtime_params_file option was used as it is currently  implemented without this PR.